lib/pbench/cli/agent/commands/tools/clear.py

```diff
@@ -25,83 +25,105 @@ def __init__(self, context):
-                status = self._clear_tools(name, remote)
-                if status > 0:
-                    errors = 1
-                elif status == 0:
-                    tool_not_found = False
-
-            tool_files = [p.name for p in tg_dir_r.iterdir()]
-
-            if len(tool_files) == 1 and tool_files[0] == "__label__":
-                label = tg_dir_r / "__label__"
-                try:
-                    label.unlink()
-                except Exception:
-                    self.logger.error("Failed to remove label for remote %s", tg_dir_r)
-                    errors = 1
-
-            if self.is_empty(tg_dir_r):
-                self.logger.info(
-                    'All tools removed from group "%s" on host "%s"',
-                    self.context.group,
-                    tg_dir_r.name,
-                )
+                    continue
+
+                if self.context.name:
+                    names = self.context.name.split(",")
+                else:
+                    # Discover all the tools registered for this remote
+                    names = self.tools(tg_dir_r)
+                    if not names:
+                        # FIXME:  this is another odd case -- the remote subdirectory
+                        #  exists, but it's empty.  (We'll remove it below.)
+                        self.logger.warn(
+                            'No tools in group "%s" on host "%s".', group, remote,
+                        )
+
+                for name in names:
+                    status = self._clear_tools(name, remote, group)
+                    if status > 0:
+                        errors = 1
+                        tools_not_found.append(name)
+                    elif status < 0:
+                        tools_not_found.append(name)
+
+                tool_files = [p.name for p in tg_dir_r.iterdir()]
+
+                if len(tool_files) == 1 and tool_files[0] == "__label__":
+                    label = tg_dir_r / "__label__"
+                    try:
+                        label.unlink()
+                    except Exception:
+                        self.logger.error(
+                            "Failed to remove label for remote %s", tg_dir_r
+                        )
+                        errors = 1
+
+                if self.is_empty(tg_dir_r):
+                    self.logger.info(
+                        'All tools removed from group "%s" on host "%s"',
+                        group,
+                        tg_dir_r.name,
+                    )
+                    try:
+                        shutil.rmtree(tg_dir_r)
+                    except OSError:
+                        self.logger.error(
+                            "Failed to remove remote directory %s", tg_dir_r
+                        )
+                        errors = 1
+
+            if self.context.name and tools_not_found:
+                self.logger.warn(f"Tools {tools_not_found} not found")
```

Comment:
Shouldn't there be a test case showing this across multiple groups? (Maybe I missed it?)

Reply:
you mean `pbench-clear-tools --group=grp1,grp2 --name=vmstat` in case if vmstat doesn't exists in one or both of these groups?

Reply:
Yes; I suppose preferably that it doesn't exist in some subset so we can see that the groups where it *does* exist work correctly despite the evident warnings in the groups where it *doesn't*.

Reply:
Yes, we should prove that we behave well when clearing a tool from multiple groups when:

- [ ] the tool is present in all the groups
- [ ] the tool is missing from one or more of the groups but is present in the others
- [ ] the tool is missing from all the groups

And, we should presumably assert the following after the clear:

- [ ] the specified tool is gone from all the groups
- [ ] no other tools have been cleared from any of the groups
- [ ] any group for which this was the only tool should have been completely removed (with the possible exception of the `default` group, but that's a separate discussion).

And, this should be the case regardless of whether the `--name` switch specifies one, two, or three tools.  😁

Reply:
Yep these use cases are covered under `test_clear_tools_test71` and `test_clear_tools_test72` test functions with all the possible assertion that Webb just pointed out.

Reply:
I have the impression that we don't test this code nearly as intensively as we ideally should.  But, I understand that there is a cost, and that we may want to make a trade off.

Adding the comma-options increases the (optimal) testing requirements combinatorially:  we now have three layers -- groups, remotes, names -- and there could be items missing from any of the layers, and we want to make sure that things are well-behaved at every layer under failures at any layer; these options can have zero values (i.e. be omitted), a value with no commas, a value with one comma, and a value with multiple commas.  So, to test this properly, it seems like we need to construct a matrix with at least 4^3 entries (more, actually).  Now, this might not actually be too hard to do, given PyTest fixtures and parameterization (we can do nested parameterization).

If we construct a fixed collection of tool groups, then we can run tests repeatedly against them.  Then we can parameterize each of the groups, remotes, and names.  Each parameterization should have cases with the option omitted, the option with a single value, the option with two values, and the option with three values; and, there should be an additional case or cases where the option references values which don't exist (optimally, the non-existent option should be tried in each of the first, second, and third positions of a multi-value option -- this would be key to testing the "tool not found" scenario which I mentioned in my review).

>  these use cases are covered under `test_clear_tools_test71` and `test_clear_tools_test72`

I don't think that's the...uh...case.  😉  '`71` tests whether the `default` tools directory is retained after a successful multi-group clear -- it doesn't do anything with remotes or individual tools; '`72` ensures that things work if a bad **group** is specified as one of the options, but, again, it doesn't touch on remotes or tools.  In the other new test cases, I think the tool is always present in all the groups/remotes.